### PR TITLE
fix(argo-server)!: Implement missing instanceID code. Fixes #2780

### LINKED ISSUE #2780: argo list does not show workflows from all instances or accept --instanceid
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [ ] I've included the logs. (N/A - can do if necessary.)
 
**What happened**:
We have deployed an argo 2.7.3 workflow controller with an instance ID in the configmap alongside a 2.3.0 controller.

`argo list` only shows workflows from the controller with no instance ID (2.3.0) and I can't work out how to get it to show the other workflows (see reproduction steps below.)

`kubectl get workflows` functions as expected.

**What you expected to happen**:

`argo list` could show all workflows, or support the `--instanceid` argument.

**How to reproduce it (as minimally and precisely as possible)**:

```
! ~ $ argo submit --instanceid 27 https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml
Name:                hello-world-c29t6
Namespace:           whatever-dev
ServiceAccount:      default
Status:              Pending
Created:             Wed Apr 22 16:18:05 +0100 (now)
! ~ $ argo submit https://raw.githubusercontent.com/argoproj/argo/master/examples/hello-world.yaml
Name:                hello-world-xbb22
Namespace:           whatever-dev
ServiceAccount:      default
Status:              Pending
Created:             Wed Apr 22 16:18:14 +0100 (now)
! ~ $ argo list | grep hello-world
hello-world-xbb22      Running     3s     3s         0
hello-world-mbg4n      Succeeded   1d     5s         0
hello-world-jpfbf      Succeeded   142d   18s        0
! ~ $ argo list --instanceid 27
Error: unknown flag: --instanceid
```

**Anything else we need to know?**:

**Environment**:
- Argo version: 2.7.3
```
$ argo version
```
- Kubernetes version :
```
v1.14.9-eks-502bfb
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

This contains a breaking change.

* When: You are using (a) instance ID and (b) workflow templates. 
* Impact: You will not longer be available access workflow templates via the Argo Server API.
* Why: You should never have been able to access this data and the vast majority of users, if any, will be impacted.

### CODE REVIEW

**Comment by sarabala1979** [comment]
The `InstanceID` concept is mainly for workflow and CronWorkflow execution purposes.  It will not apply on definitions of WorkflowTemplate and ClusterWorkflowTemplate.  It is a simple flexibility to control the workflow execution particular controller.   @jessesuen you can comment on this.

**Comment by alexec** [comment]
@jessesuen @sarabala1979 @dtaniwaki there are a lot of misses in the code base for instance ID. This tells me that is fundamentally hard to get right - like RBAC.

I essentially had to intercept every single call. Really hard to test.

It made me wonder if it is possible to do this in the generated code? @alexmt

File: server/clusterworkflowtemplate/cluster_workflow_template_server.go [hunk 1]
```diff
@@ -39,15 +43,24 @@ func (cwts *ClusterWorkflowTemplateServer) CreateClusterWorkflowTemplate(ctx con
...
 	if err != nil {
 		return nil, err
 	}
+	return wfTmpl, nil
+}
 
-	return wfTmpl, err
+func (cwts *ClusterWorkflowTemplateServer) getTemplateAndValidate(ctx context.Context, name string) (*v1alpha1.ClusterWorkflowTemplate, error) {
+	wfClient := auth.GetWfClient(ctx)
+	wfTmpl, err := wfClient.ArgoprojV1alpha1().ClusterWorkflowTemplates().Get(name, v1.GetOptions{})
```
Comment (sarabala1979) [inline server/clusterworkflowtemplate/cluster_workflow_template_server.go:55]: See my comments for instanceID validation

File: server/clusterworkflowtemplate/cluster_workflow_template_server.go [hunk 2]
```diff
@@ -68,8 +81,11 @@ func (cwts *ClusterWorkflowTemplateServer) ListClusterWorkflowTemplates(ctx cont
 
 func (cwts *ClusterWorkflowTemplateServer) DeleteClusterWorkflowTemplate(ctx context.Context, req *clusterwftmplpkg.ClusterWorkflowTemplateDeleteRequest) (*clusterwftmplpkg.ClusterWorkflowTemplateDeleteResponse, error) {
 	wfClient := auth.GetWfClient(ctx)
-
-	err := wfClient.ArgoprojV1alpha1().ClusterWorkflowTemplates().Delete(req.Name, &v1.DeleteOptions{})
+	_, err := cwts.getTemplateAndValidate(ctx, req.Name)
```
Comment (sarabala1979) [inline server/clusterworkflowtemplate/cluster_workflow_template_server.go:84]: See my comments for instanceID validation

**Comment by sarabala1979** [comment]
This PR will change the current understanding/assumption of Argo usage with `InstanceID`. 
1. User should have separate ApiServer/UI/Controller for each InstanceID
2. User needs to configure ENV or pass the `-Argo-server URL` for each InstanceID otherwise they need to use classic client
This change needs to be documented and demoed in a community meetings.

**Comment by alexec** [comment]
> This PR will change the current understanding/assumption of Argo usage with `InstanceID`.
> 
> 1. User should have separate ApiServer/UI/Controller for each InstanceID
> 2. User needs to configure ENV or pass the `-Argo-server URL` for each InstanceID otherwise they need to use classic client
>    This change needs to be documented and demoed in a community meetings.

I think I'm fine with that. My take (and I believe @jessesuen 's too) is that instance ID is a data segregation feature and this is how it should behave. We should warn people in release notes too.

**Comment by sarabala1979** [comment]
InstanceID checking is missing in the Workflow Archive module.

**Comment by alexec** [comment]
> InstanceID checking is missing in the Workflow Archive module.

This is enforced by `workflow_archive.go` - which is not the same as others.

File: server/workflow/workflow_server.go [hunk 3]
```diff
@@ -205,11 +191,10 @@ func (s *workflowServer) WatchWorkflows(req *workflowpkg.WatchWorkflowsRequest,
 }
 
 func (s *workflowServer) DeleteWorkflow(ctx context.Context, req *workflowpkg.WorkflowDeleteRequest) (*workflowpkg.WorkflowDeleteResponse, error) {
-	_, err := s.getWorkflow(ctx, req.Namespace, req.Name, metav1.GetOptions{})
+	_, err := s.getWorkflowAndValidate(ctx, req.Namespace, req.Name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (alexec) [reply server/workflow/workflow_server.go:194]: in the `GetOptions`?

[the remark this answers is not in the document]
Comment (alexec) [reply server/workflow/workflow_server.go:194]: `GetOptions` does not support a label filter, nor does `DeleteOptions` only `ListOptions`.

[the remark this answers is not in the document]
Comment (sarabala1979) [reply server/workflow/workflow_server.go:192]: got it, you are right. thanks

**Comment by alexec** [comment]
Failed builds indicate dynamic check is working - but I've missed something :(

**Comment by alexec** [comment]
@sarabala1979 this is ready for review again. 
@jessesuen there were bugs in the proto files - the `name` field is ignored - can you approve please?

**Comment by alexec** [comment]
@sarabala1979 are you still reviewing please?

**Pull request merged**